subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+	size_t ate_size;
+	/* Initialize addr to 0 for the case fs->sector_count == 0. This
+	 * should never happen as this is verified in zms_mount() but both
+	 * Coverity and GCC believe the contrary.
+	 */
+	uint64_t addr = 0U, data_wra = 0U;
+	uint32_t i, closed_sectors = 0;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	/* step through the sectors to find a open sector following
+	 * a closed sector, this is where zms can write.
+	 */
+
+	for (i = 0; i < fs->sector_count; i++) {
+		addr = ((uint64_t)i << ADDR_SECT_SHIFT) + (uint32_t)(fs->sector_size - ate_size);
+
+		/* read the first and second ate to find out if sector is
+		 * open/close
+		 */
+		rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+		if (rc) {
+			goto end;
+		}
+
+		/* update cycle count */
+		fs->sector_cycle = empty_ate.cycle_cnt;
+
+		if (zms_close_ate_valid(fs, &close_ate) && zms_empty_ate_valid(fs, &empty_ate) &&
+		    (close_ate.cycle_cnt == empty_ate.cycle_cnt)) {
+			/* closed sector */
+			closed_sectors++;
+			/* Let's check first that we support this ZMS version */
+			if (ZMS_GET_VERSION(empty_ate.metadata) != ZMS_DEFAULT_VERSION) {
+				LOG_ERR("ZMS Version is not supported");
+				rc = -ENOEXEC;
+				goto end;
+			}
+			zms_sector_advance(fs, &addr);
+			/* addr is pointing to the close ATE */
+			rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+			if (rc) {
+				goto end;
+			}
+			/* update cycle count */
+			fs->sector_cycle = empty_ate.cycle_cnt;
+			if (!zms_empty_ate_valid(fs, &empty_ate) ||
+			    !zms_close_ate_valid(fs, &close_ate) ||
+			    (close_ate.cycle_cnt != empty_ate.cycle_cnt)) {
+				/* open sector */
+				break;
+			}
+		}
+	}
+	/* all sectors are closed, this is not a zms fs */
+	if (closed_sectors == fs->sector_count) {
+		rc = -EDEADLK;
+		goto end;
+	}
+
+	if (i == fs->sector_count) {
+		/* none of the sectors were closed, which means that the first
+		 * sector is the one in use, except if there are only 2 sectors.
+		 * Let's check if the last sector has valid ATEs otherwise set
+		 * the open sector to the first one.
+		 */
+		rc = zms_flash_ate_rd(fs, addr - 2 * ate_size, &first_ate);
+		if (rc) {
+			goto end;
+		}
+		if (!zms_ate_valid(fs, &first_ate)) {
+			zms_sector_advance(fs, &addr);
+		}
+		rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+		if (rc) {
+			goto end;
+		}
+		if (!zms_empty_ate_valid(fs, &empty_ate)) {
+			if (ebw_required) {
+				rc = zms_flash_erase_sector(fs, addr);
+				if (rc) {
+					goto end;
+				}
+			}
+			rc = zms_add_empty_ate(fs, addr);
+			if (rc) {
+				goto end;
+			}
+		} else {
+			/* empty_ate is valid let's check the version */
+			if (ZMS_GET_VERSION(empty_ate.metadata) != ZMS_DEFAULT_VERSION) {
+				LOG_ERR("ZMS Version is not supported");
+				rc = -ENOEXEC;
+				goto end;
+			}
```

Comment:
Reverse the blocks: expect valid first.
Move the block in `!zms_empty_ate_valid` to some function like "zms_prepare/zms_format" to make it clear what is happening and move the 1053-1054 comment to the block.

Reply:
Shouldn't we also parametrize the operation so that user could, either via parameter to mount or context,
select to:
 - do nothing if no valid system is found, except returning error
 - enforce re-initialization of system?